### PR TITLE
feat: text shape

### PR DESCRIPTION
**Pull request by NickKhalow**
Component
- Based on TextShapeRendererComponent, the component incapsulates ITextShapeRenderer that provides rendering for text

ITextShapeRenderer
- Supports Hide/Show functionality
- Supports applying properties from ProtoBuf component PBTextShape

Factory
- Provides ITextShapeRenderer uncoupled with implementation

TMPTextShapeRenderer
- Implements ITextShapeRenderer
- Built upon TextMeshPro with MeshRenderer
- Uses MaterialPropertyBlock to avoid memory allocation on instantiating new materials

Fonts
- Assign font TMP files in the setting and reuse it with Dynamic World

Systems
- Instantiation with budgeting to support stable performance 
- Support for VisibilityComponent

Demo World
- Explore the component in _Assets/DCL/SDKComponents/TextShape/Demo/TextShapePlayground.unity_ and tweak any parameter
- Several Demo World to combine already existing world without copy/paste
- Tested with BillboardComponent

### CODE REVIEW

File: Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using System;
+
+namespace DCL.DemoWorlds.Extensions
+{
+    public static class NullExtensions
+    {
+        public static T EnsureNotNull<T>(this T? value, string? message = null)
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs:7]: Does this extension need to be moved to a more genralised extension location? It isn't fully related to DemoWorlds? Does it need constraints for references only?

File: Explorer/Assets/DCL/SDKComponents/TextShape/Component/TextShapeRendererComponent.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            if (visibility)
+                textShape.Show();
+            else
+                textShape.Hide();
+        }
+
+        public void Dispose()
+        {
+            Debug.LogWarning("Dispose logic is not implemented");
+            //todo
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/SDKComponents/TextShape/Component/TextShapeRendererComponent.cs:34]: Does this need implementing?

File: Explorer/Assets/DCL/SDKComponents/TextShape/Renderer/Factory/PoolTextShapeRendererFactory.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+using DCL.Optimization.Pools;
+using DCL.SDKComponents.TextShape.Fonts;
+using UnityEngine;
+
+namespace DCL.SDKComponents.TextShape.Renderer.Factory
+{
+    /// <summary>
+    /// Is not thread safe
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/SDKComponents/TextShape/Renderer/Factory/PoolTextShapeRendererFactory.cs:8]: We are using ThreadSafeComponentPool here, is it unsafe? Can this factory be used on multiple threads?

File: Explorer/Assets/DCL/SDKComponents/TextShape/System/InstantiateTextShapeSystem.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        public InstantiateTextShapeSystem(World world, ITextShapeRendererFactory textShapeRendererFactory, IConcurrentBudgetProvider instantiationFrameTimeBudgetProvider) : base(world)
+        {
+            this.textShapeRendererFactory = textShapeRendererFactory;
+            this.instantiationFrameTimeBudgetProvider = instantiationFrameTimeBudgetProvider;
+        }
+
+        public InstantiateTextShapeSystem(World world, ITextShapeRendererFactory textShapeRendererFactory) : this(
+            world,
+            textShapeRendererFactory,
+            new FrameTimeCapBudgetProvider(
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/SDKComponents/TextShape/System/InstantiateTextShapeSystem.cs:32]: Isn't 33ms most of the timeframe? Should this be lower?

File: Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs (hunk 1 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs:7]: Good suggestion, where is the general extension assembly?

File: Explorer/Assets/DCL/SDKComponents/TextShape/Renderer/Factory/PoolTextShapeRendererFactory.cs (hunk 3 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/SDKComponents/TextShape/Renderer/Factory/PoolTextShapeRendererFactory.cs:8]: I saw comments above ThreadSafeObjectPool<T> in DCL.Optimization.Pools.ComponentPool<T> that it's not thread-safe. So it's need to be clarified. I haven't tests now that covers multithreading use-cases

```
/// <summary>
///     It is not thread-safe
/// </summary>
private readonly ThreadSafeObjectPool<T> objectPool;
```

File: Explorer/Assets/DCL/SDKComponents/TextShape/System/InstantiateTextShapeSystem.cs (hunk 4 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/SDKComponents/TextShape/System/InstantiateTextShapeSystem.cs:32]: I took this value from DCL.PluginSystem.Global.StaticSettings as default

```
// Performance budgeting
[field: Header("Performance Budgeting")] [field: Space]
[field: SerializeField]
public int FrameTimeCap { get; private set; } = 33; // in [ms]. Table: 33ms ~ 30fps | 16ms ~ 60fps | 11ms ~ 90 fps | 8ms ~ 120fps
```

File: Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs (hunk 1 above)
Comment (m3taphysics) [reply Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs:7]: You may have to create it - I don't know!

File: Explorer/Assets/DCL/SDKComponents/TextShape/Component/TextShapeRendererComponent.cs (hunk 2 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/SDKComponents/TextShape/Component/TextShapeRendererComponent.cs]: No, I pushed changes

File: Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs (hunk 1 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/DemoWorlds/Extensions/NullExtensions.cs:7]: So I'll ask in chats to the team, maybe we already have one

**Review by m3taphysics — APPROVED** [review]

**Pull request merged**